worlds/sa2b/Rules.py

```diff
@@ -2,352 +2,856 @@
 
 from BaseClasses import MultiWorld
 from .Names import LocationName, ItemName
-from .Locations import first_mission_location_table, second_mission_location_table, third_mission_location_table, \
-    fourth_mission_location_table, fifth_mission_location_table, \
-    upgrade_location_table, boss_gate_set
+from .Locations import boss_gate_set
 from ..AutoWorld import LogicMixin
-from ..generic.Rules import add_rule, set_rule
+from ..generic.Rules import add_rule, set_rule, CollectionRule
 from .GateBosses import boss_has_requirement
+from .Missions import stage_name_prefixes, mission_orders
 
 
-def set_mission_progress_rules(world: MultiWorld, player: int):
-    for (k1, v1), (k2, v2), (k3, v3), (k4, v4), (k5, v5) in \
-            zip(sorted(first_mission_location_table.items()),
-                sorted(second_mission_location_table.items()),
-                sorted(third_mission_location_table.items()),
-                sorted(fourth_mission_location_table.items()),
-                sorted(fifth_mission_location_table.items())):
+def add_rule_safe(multiworld: MultiWorld, spot_name: str, player: int, rule: CollectionRule):
+    try:
+        add_rule(multiworld.get_location(spot_name, player), rule)
+    except KeyError:
+        # Do nothing for mission locations that do not exist
+        return
```

Comment:
```suggestion
    try:
        location = multiworld.get_location(spot_name, player)
    except KeyError:
        # Do nothing for mission locations that do not exist
        pass
    else:
        add_rule(location, rule)
```
Generally a good idea to keep the try as small as possible to not catch other unrelated errors, especially when silently ignoring the error.